// edgedb/dist/src/index.shared.js
"use strict";
/*!
 * This source file is part of the EdgeDB open source project.
 *
 * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var datetime_1 = require("./datatypes/datetime");
exports.LocalDateTime = datetime_1.LocalDateTime;
exports.LocalDate = datetime_1.LocalDate;
exports.LocalTime = datetime_1.LocalTime;
exports.Duration = datetime_1.Duration;
var set_1 = require("./datatypes/set");
exports.Set = set_1.Set;
var uuid_1 = require("./datatypes/uuid");
exports.UUID = uuid_1.UUID;
__export(require("./errors"));
const reg = require("./codecs/registry");
const buf = require("./buffer");
const introspect = require("./datatypes/introspect");
exports._CodecsRegistry = reg.CodecsRegistry;
exports._ReadBuffer = buf.ReadBuffer;
exports._introspect = introspect.introspect;
const bigint_1 = require("./bigint");
exports._plugJSBI = bigint_1.plugJSBI;
